dali/daliadmin/daliadmin.cpp

```diff
@@ -691,6 +692,104 @@ static int clusterGroup(const char *name, const char *outputFilename)
+    StringBuffer xpath;
+    xpath.append("*[@name='").append(name).append("']");
+    Owned<IPropertyTree> match = root->getPropTree(xpath);
+    if (match)
+        return match.getClear();
+    ERRLOG("Path %s not found", name);
+    return nullptr;
+}
+
+static IPropertyTree * selectPath(IPropertyTree * root, const char * path)
+{
+    if (!path || !*path)    // use / to refer to the root directory
+        return LINK(root);
+
+    const char * split = strstr(path, "::");
+    if (split)
+    {
+        //Can use :: to refer to the root directory
+        if (split == path)
+            return selectPath(root, split + 2);
+
+        StringAttr name(path, split - path);
+        Owned<IPropertyTree> match = selectLevel(root, name);
+        if (match)
+            return selectPath(match, split + 2);
+        return nullptr;
+    }
+    return selectLevel(root, path);
+}
+
+static void displayDirectory(IPropertyTree * directory, const char * options, unsigned depth)
+{
+    Owned<IPropertyTreeIterator> elems = directory->getElements("*");
+    ForEach(*elems)
+    {
+        IPropertyTree & cur = elems->query();
+        const char * tag = cur.queryName();
+        const char * name = cur.queryProp("@name");
+        const char * modified = cur.queryProp("@modified");
+        if (name && tag)
+        {
+            if (strieq(tag, "Scope"))
+            {
+                OUTLOG("%*sD %s", depth, "", name);
+                if (options && strchr(options, 'r'))
+                    displayDirectory(&cur, options, depth+1);
+            }
+            else if (strieq(tag, "File"))
+            {
+                const char * group = cur.queryProp("@group");
+                const char * size = cur.queryProp("Attr[1]/@size");
+                if (options && strchr(options, 'l'))
+                    OUTLOG("%*s  %-30s %12s %s %s", depth, "", name, size ? size : "", group ? group : "?", modified ? modified : "");
+                else
+                    OUTLOG("%*s  %s", depth, "", name);
+            }
+            else if (strieq(tag, "SuperFile"))
+            {
+                if (options && strchr(options, 'l'))
+                    OUTLOG("%*sS %s %s (%d)", depth, "", name, modified ? modified : "", cur.getPropInt("@numsubfiles"));
+                else
+                    OUTLOG("%*sS %s", depth, "", name);
+
+                if (options && strchr(options, 's'))
+                {
+                    Owned<IPropertyTreeIterator> subs = cur.getElements("SubFile");
+                    ForEach(*subs)
+                    {
+                        OUTLOG("%*s->%s", depth, "", subs->query().queryProp("@name"));
+                    }
+                }
+            }
+            else
+                OUTLOG("? %s %s", name, tag);
+        }
+    }
+}
+
+static void dfsLs(const char *name, const char *options, bool safe = false)
+{
+    StringBuffer xpath;
+    Owned<IRemoteConnection> conn = connectXPathOrFile("/Files",safe,xpath);
```

Comment:
A bit odd to use connectXPathOrFile, when in this case it is statically a xpath ("/Files").
Would be clearer if used connect() directly, i.e.:

Owned<IRemoteConnection> conn = querySDS().connect("/Files", myProcessSession(), 0, daliConnectTimeoutMs);

Not new, but the use of 'safe' in connectXPathOrFile (and hence the dfsLs passed param) is odd.
If safe then avoids locking.
However, it will always mark safe = true in this case, as it (connectXPathOrFile) determines /Files is safe because it is a root level node (I am not sure that makes a great deal of sense).

I suspect it would be better to avoid locking altogether for dfsls.